### ISSUE TITLE
Add `unused_argument` lint

### ISSUE THREAD

**Issue by cubuspl42** [dart-lang/linter#2204]
Add a lint for detecting unused function arguments, together with a helper function for marking arguments unused on purpose (for example for maintaining backward compatible public interface). This pair would be analogous to [`unawaited_futures`](https://dart-lang.github.io/linter/lints/unawaited_futures.html) and [`unawaited`](https://pub.dev/documentation/pedantic/latest/pedantic/unawaited.html).

**Examples**

```dart
void f(int a) {
}

// info: The value of the argument 'a' isn't used. (unused_argument at ...)
```

```dart
void f(int a) {
  unused(a);
}

// everything is fine; no analyzer output
```

**Comment by bwilkerson** [comment]
There are two other ways that I can think of to silence the lint, so it isn't clear to me that `unused` adds any value. One is to use an `ignore` comment, the other would be to add an explicit reference such as `assert(a != null);` (which, by the way, is exactly what `unused` does but without the runtime overhead).

@srawlins who has recently done some work in this area

**Comment by srawlins** [comment]
I think there are several reasons we haven't written any analysis for unused parameters:

1. the potential for a widely scoped lint to result in many false positives
2. the non-local difficulty in removing the lint
3. a lint rule without false positives would be incredibly narrowly scoped

Take this example:

```dart
class _C {
  int m(double x, int y) => y;
}
```

While _C is private, it could be sub-classed in the same library by a public class, and that public class is now open to sub-classing outside the library. So two things can happen outside this library (even outside the package):

* subclasses might override `_C.m`, so that removing the first parameter would invalidate those overrides. You could still remove the parameter, but it becomes a breaking change.
* Callers to `_C.m` would need to be updated. Again, even with a public method of a private class, it could be called outside the library.
* If `_C.m` overrides a method in a super class, the developer _cannot_ remove the `x` parameter without changing the signature of the overridden method, which may be outside their ownership.
* If `_C.m` is ever torn-off, it's signature becomes important in hard-to-track places. This can also happen outside the library, as an instance of `_C` can escape the library, and its `m` method can be torn-off.

So if you want to avoid reporting lint to users that require breaking changes, and changes to the function's callers, subclasses, or superclasses, or `// ignore` comments on methods that override, you need to narrow the scope to a short list of functions:

* private static methods of classes (that are never torn-off),
* private top-level functions (that are never torn-off),
* methods on private extensions (that are never torn-off),
* end of list? I'm probably missing something.

It might be worth the effort to write both a zero-false-positive Hint for the list above, and one or two wider-reaching Lints for methods which override, can be overridden, or otherwise escape the library.

**Comment by srawlins** [comment]
Also, @matanlurey 's suggestions for unused_construct_parameters apply here:

* Maybe don't complain about unused parameters named `_`
* Maybe don't complain about unused `@deprecated` parameters.

**Comment by rrousselGit** [comment]
+1 on using `_` to explicitly tell the linter that the variable is voluntarily unused.

```dart
void fn(int _, String __, int bar) {
  // `bar` is unused. But no warning for _ and __
}
```

**Comment by jngbng** [comment]
Another approach is not to complain about parameters that starts with `_`, which is adopted by some lints.

```
void fn(int _foo, String _msg, int bar) {
  // But no warning for _foo and _msg
}
```

**Comment by cubuspl42** [comment]
Hey, any progress on this? Really, there are many cases when we could already have this without any need for manual overrides, for example:

- Library-private methods
- Methods in a library-private class which are known not to be overriden within this library
- Library-private top-level functions
- Methods annotated with `@nonVirtual` (as I can't see any reasons why we can't depend on annotations from `meta`)

I won't even mention what we could do assuming a possibility of whole-application analysis, but first things first.

**Comment by incendial** [comment]
If you're willing to try DCM plugin, this rule is available https://dcm.dev/docs/individuals/rules/common/avoid-unused-parameters/ for more then a year.
